styx-common/src/main/java/com/spotify/styx/util/ShardedCounter.java

```diff
@@ -0,0 +1,339 @@
+   */
+  private CounterSnapshot refreshCounterSnapshot(String counterId) {
+    final CounterSnapshot newSnapshot = new CounterSnapshot(datastore, counterId, time);
+    cache.put(counterId, newSnapshot);
+    return newSnapshot;
+  }
+
+  /**
+   * Must be called within a TransactionCallable. (?)
+   *
+   * <p>Augments the transaction with operations to persist the given limit in Datastore. So long as
+   * there has been no preceding successful updateLimit operation, no limit is applied in
+   * updateCounter operations on this counter.
+   */
+  public void updateLimit(DatastoreReaderWriter transaction, String counterId, long limit) {
+    final Key limitKey = datastore.newKeyFactory().setKind(KIND_COUNTER_LIMIT).newKey(counterId);
+
+    transaction.put(Entity.newBuilder(limitKey).set(PROPERTY_LIMIT, limit).build());
+  }
+
+  /**
+   * Reads the latest limit value from Datastore, for the specified {@param counterId}
+   */
+  public static long getLimit(Datastore datastore, String counterId) {
+    final Key limitKey = datastore.newKeyFactory().setKind(KIND_COUNTER_LIMIT).newKey(counterId);
+    final Entity limitEntity = datastore.get(limitKey);
+    if (limitEntity == null) {
+      return Long.MAX_VALUE;
+      // Or IllegalStateException("No limit found in Datastore for " + counterId);?
+    } else {
+      return limitEntity.getLong(PROPERTY_LIMIT);
+    }
+  }
+
+  /**
+   * Must be called within a TransactionCallable. Augments the transaction with certain operations
+   * that strongly consistently increment resp. decrement the counter referred to by counterId, and
+   * cause the transaction to fail to commit if the counter's associated limit is exceeded. Also
+   * spurious failures are possible.
+   *
+   * <p>Delta should be +/-1 for graceful behavior, due to how sharding is currently implemented.
+   * Updates with a larger delta are prone to spuriously fail even when the counter is not near to
+   * exceeding its limit. Failures are certain when delta >= limit / NUM_SHARDS + 1.
+   */
+  public void updateCounter(DatastoreReaderWriter transaction, String counterId, long delta) {
+    CounterSnapshot snapshot = getCounterSnapshot(counterId);
+    int shardIndex = snapshot.pickShardWithSpareCapacity(delta);
+
+    final Key shardKey = datastore.newKeyFactory().setKind(KIND_COUNTER_SHARD)
+        .newKey(counterId + "-" + shardIndex);
+    final Entity shard = transaction.get(shardKey);
+
+    if (shard != null) {
+      final long newShardValue = shard.getLong(PROPERTY_SHARD_VALUE) + delta;
+      if (Range.closed(0L, snapshot.shardCapacity(shardIndex))
+          .contains(newShardValue)) {
+        transaction.put(Entity.newBuilder(shard)
+                            .set(PROPERTY_SHARD_VALUE,
+                                 shard.getLong(PROPERTY_SHARD_VALUE) + delta)
+                            .build());
+      } else {
+        // TODO fail (rollback) the transaction
+        throw new ConcurrentModificationException(
```

Comment:
I mean, this here should be the place we throw `CounterCapacityException` (and make sure that it's handled by styx in a way that rolls back the entire transaction).

And if we avoid throwing it in multiple places, then we have less code and more consistent execution paths.

Reply:
I used ConcurrentModificationException just as a placeholder, the CounterCapacityException is a better class for it.

And maybe we'll want it to inherit from TransactionException (from https://github.com/spotify/styx/pull/363) when we try to integrate the PRs...

Reply:
Or if `StorageTransaction` has a better way to signal need to abort, then we could use that instead of throwing exceptions, but there is none yet.